[Drill][Added] CSV Drill Table output

nguyen-v opened the pull request:
This PR adds a CSV drill table output to the `excellon` output. I've added a simple test case in `test_drill.py`.
It's basically a reimplementation of
https://gitlab.com/kicad/code/kicad/-/blob/master/pcbnew/exporters/gendrill_file_writer_base.cpp
It should work for KiCad 5+, but for KiCad 5 the generated table is slightly different (because of missing attributes).
The column order/names can be changed similar to how it's done in bom.
Example from `drill.kibot.yaml`
```
# Drills and Gerber drills
kibot:
  version: 1

outputs:

  - name: excellon_drill
    comment: "Excellon drill files"
    type: excellon
    dir: Drill
    options:
      metric_units: true
      pth_and_npth_single_file: false
      use_aux_axis_as_origin: false
      minimal_header: false
      mirror_y_axis: false
      report: '%f-%i.%x'
      map: 'pdf'
      table:
        pth_and_npth_single_file: false

  - name: gerber_drills
    comment: "Gerber drill files"
    type: gerb_drill
    dir: Drill
    options:
      use_aux_axis_as_origin: false
      map: 'pdf'
```

I've also modified the 3Rs_bv.kicad_pcb PCB for KiCad 5, as it was not similar to the other versions (there was a missing through hole via, and the tests I wrote are expecting a similar PCB for all versions)

nguyen-v commented:
I'm not very satisfied with having to specify `pth_and_npth_single_file` twice. Should I move the `DrillTable` class to `out_excellon.py` instead of `out_any_drill.py`?

set-soft commented:
> I'm not very satisfied with having to specify `pth_and_npth_single_file` twice. Should I move the `DrillTable` class to `out_excellon.py` instead of `out_any_drill.py`?

If the user wants gerber drills and drill table then DrillTable is for both and goes to `out_any_drill.py`
If the user wants a table for PTH and another for NPTH, but just one drill file for the drill machine then we need 2 options.
We could change the DrillTable version to be named differently, lets say "unify_pth_and_npth" and have values 'yes', 'no' and 'auto'. With 'auto' as default. Then 'auto' can copy the excellon `pth_and_npth_single_file` and be 'no' for gerbers.

set-soft commented:
Oh! BTW, the layer numbering is changing on KiCad 9 so we can't rely on B_Cu in this way.
Now F_Cu is 0, B_Cu is 2 and the inner layers are 4, 6, ... 62

nguyen-v commented:
> We could change the DrillTable version to be named differently, lets say "unify_pth_and_npth" and have values 'yes', 'no' and 'auto'. With 'auto' as default. Then 'auto' can copy the excellon pth_and_npth_single_file and be 'no' for gerbers.

Sounds good

> Oh! BTW, the layer numbering is changing on KiCad 9 so we can't rely on B_Cu in this way. Now F_Cu is 0, B_Cu is 2 and the inner layers are 4, 6, ... 62

https://gitlab.com/kicad/code/kicad/-/commit/a4292ea51631001c60cb01ac759606a13747f861
I will have a look at it

nguyen-v commented:
I've changed `pth_and_npth_single_file` to `unify_pth_and_npth` for the `DrillOptions`. By default it will be `auto`. If it detects that there is a `pth_and_npth_single_file` in the class (i.e. `ExcellonOptions` or any derivated class), it will use the `pth_and_npth_single_file` if `unify_pth_and_npth = 'auto'` else it will just use `unify_pth_and_npth`.
For classes without `pth_and_npth_single_file` (i.e. derivatives of `AnyDrill`), `unify_pth_and_npth` will default to 'no' if nothing is specified. 
I've added some support for the new layer numbering (I think it only affects the function that gets the layer names in physical order, front being 1 and back being the # of layers). Could you have a quick look at it?

set-soft commented:
> Could you have a quick look at it?

Look ok.

Also take a look to the review about `GS.get_modules()`

nguyen-v commented:
> Also take a look to the review about `GS.get_modules()`

What do you mean? Did you write a review somewhere? Is it about using `GS.get_modules()` here https://github.com/nguyen-v/KiBot/blob/c6241a02748a94ff6ee2aedc554267303e54c14d/kibot/kicad/drill_info.py#L193

set-soft commented:
> > Also take a look to the review about `GS.get_modules()`
> 
> What do you mean? Did you write a review somewhere? Is it about using `GS.get_modules()` here https://github.com/nguyen-v/KiBot/blob/c6241a02748a94ff6ee2aedc554267303e54c14d/kibot/kicad/drill_info.py#L193

Yes, this should be a call to GS.get_modules(), which already does what you wrote there.

set-soft commented:
Now should be visible, I forgot to mark it as ready

nguyen-v commented:
> Hi @nguyen-v ! I did some changes:
> 
> 1. I removed spaces in the layer pair names, spaces in file names are a bad thing
> 2. I changed the way columns are validated using more internal checks instead of manual checks

Thanks!
